File: internal/ws/token.go
package ws

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"sync"
	"time"
)

const (
	tokenBytes  = 32
	tokenTTL    = 5 * time.Minute
	cleanupTick = 60 * time.Second
)

// tokenEntry holds a token along with its creation time for TTL enforcement.
type tokenEntry struct {
	teamID    string
	createdAt time.Time
}

// TokenManager generates and validates one-time tokens for WebSocket authentication.
// Tokens are stored in-memory only and consumed on first use.
// Tokens expire after 5 minutes even if not consumed.
type TokenManager struct {
	// tokens maps token string -> tokenEntry
	tokens map[string]tokenEntry
	mu     sync.Mutex
	stopCh chan struct{}
	once   sync.Once
}

// NewTokenManager creates a new TokenManager and starts the background cleanup goroutine.
func NewTokenManager() *TokenManager {
	tm := &TokenManager{
		tokens: make(map[string]tokenEntry),
		stopCh: make(chan struct{}),
	}
	tm.once.Do(func() {
		go tm.cleanupLoop()
	})
	return tm
}

// GenerateToken creates a cryptographically random one-time token associated
// with the given team ID. The token is 32 bytes, hex encoded (64 chars).
// Tokens expire after 5 minutes.
func (tm *TokenManager) GenerateToken(teamID string) (string, error) {
	b := make([]byte, tokenBytes)
	if _, err := rand.Read(b); err != nil {
		return "", fmt.Errorf("failed to generate token: %w", err)
	}

	token := hex.EncodeToString(b)

	tm.mu.Lock()
	tm.tokens[token] = tokenEntry{teamID: teamID, createdAt: time.Now()}
	tm.mu.Unlock()

	return token, nil
}

// ValidateAndConsume atomically checks if a token is valid and removes it.
// Returns the team ID and true if valid, or empty string and false if invalid/consumed/expired.
func (tm *TokenManager) ValidateAndConsume(token string) (string, bool) {
	tm.mu.Lock()
	defer tm.mu.Unlock()

	entry, ok := tm.tokens[token]
	if !ok {
		return "", false
	}

	if time.Since(entry.createdAt) > tokenTTL {
		delete(tm.tokens, token)
		return "", false
	}

	delete(tm.tokens, token)
	return entry.teamID, true
}

// Validate checks if a token is valid without consuming it.
// Returns the team ID and true if valid and not expired.
func (tm *TokenManager) Validate(token string) (string, bool) {
	tm.mu.Lock()
	defer tm.mu.Unlock()

	entry, ok := tm.tokens[token]
	if !ok {
		return "", false
	}

	if time.Since(entry.createdAt) > tokenTTL {
		delete(tm.tokens, token)
		return "", false
	}

	return entry.teamID, true
}

// Consume removes a previously validated token. Returns false if already consumed or expired.
func (tm *TokenManager) Consume(token string) bool {
	tm.mu.Lock()
	defer tm.mu.Unlock()

	entry, ok := tm.tokens[token]
	if !ok {
		return false
	}

	if time.Since(entry.createdAt) > tokenTTL {
		delete(tm.tokens, token)
		return false
	}

	delete(tm.tokens, token)
	return true
}

// PendingCount returns the number of unexpired, unused tokens. Useful for testing.
func (tm *TokenManager) PendingCount() int {
	tm.mu.Lock()
	defer tm.mu.Unlock()
	now := time.Now()
	count := 0
	for _, entry := range tm.tokens {
		if now.Sub(entry.createdAt) <= tokenTTL {
			count++
		}
	}
	return count
}

// InjectExpiredToken inserts an already-expired token into the manager.
// Used only in tests — do not call from production code.
func (tm *TokenManager) InjectExpiredToken(token, teamID string) {
	tm.mu.Lock()
	defer tm.mu.Unlock()
	tm.tokens[token] = tokenEntry{teamID: teamID, createdAt: time.Now().Add(-2 * tokenTTL)}
}

// Close stops the background cleanup goroutine.
func (tm *TokenManager) Close() {
	select {
	case <-tm.stopCh:
		// Already closed
	default:
		close(tm.stopCh)
	}
}

// cleanupLoop periodically removes expired tokens to prevent unbounded growth.
func (tm *TokenManager) cleanupLoop() {
	ticker := time.NewTicker(cleanupTick)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			tm.cleanupExpiredTokens()
		case <-tm.stopCh:
			return
		}
	}
}

// cleanupExpiredTokens removes all tokens older than the TTL.
func (tm *TokenManager) cleanupExpiredTokens() {
	tm.mu.Lock()
	defer tm.mu.Unlock()

	now := time.Now()
	for token, entry := range tm.tokens {
		if now.Sub(entry.createdAt) > tokenTTL {
			delete(tm.tokens, token)
		}
	}
}

File: backend/src/ws/token.ts
/**
 * OpenHive Backend - WebSocket One-Time Token Manager
 *
 * Generates cryptographically random one-time tokens for WebSocket
 * authentication. Tokens are:
 *   - 32 bytes of entropy, hex-encoded to 64 characters
 *   - Consumed on first use (validate-then-delete)
 *   - Expired after TOKEN_TTL_MS (5 minutes)
 *   - Cleaned up by a background interval every CLEANUP_INTERVAL_MS (60 seconds)
 */

import { randomBytes } from 'node:crypto';

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/** Number of random bytes per token (results in a 64-char hex string). */
const TOKEN_BYTES = 32;

/** Token time-to-live in milliseconds (5 minutes). */
const TOKEN_TTL_MS = 5 * 60 * 1000;

/** How often the cleanup loop removes expired tokens (60 seconds). */
const CLEANUP_INTERVAL_MS = 60 * 1000;

// ---------------------------------------------------------------------------
// Internal types
// ---------------------------------------------------------------------------

/** Holds a pending token's associated team and creation time. */
interface TokenEntry {
  teamId: string;
  createdAt: Date;
}

// ---------------------------------------------------------------------------
// TokenManager
// ---------------------------------------------------------------------------

/**
 * Generates and validates one-time tokens for WebSocket authentication.
 * Tokens are stored in-memory and consumed on first use.
 * Tokens expire after TOKEN_TTL_MS even if not consumed.
 */
export class TokenManager {
  private readonly tokens: Map<string, TokenEntry> = new Map();
  private readonly cleanupInterval: ReturnType<typeof setInterval>;

  constructor() {
    this.cleanupInterval = setInterval(() => {
      this.cleanupExpiredTokens();
    }, CLEANUP_INTERVAL_MS);

    // Allow the Node.js process to exit even if the interval is still running.
    if (typeof this.cleanupInterval.unref === 'function') {
      this.cleanupInterval.unref();
    }
  }

  /**
   * Generates a cryptographically random one-time token associated with the
   * given team ID. The token is TOKEN_BYTES bytes, hex-encoded (64 chars).
   * Tokens expire after TOKEN_TTL_MS.
   */
  generateToken(teamId: string): string {
    const buf = randomBytes(TOKEN_BYTES);
    const token = buf.toString('hex');
    this.tokens.set(token, { teamId, createdAt: new Date() });
    return token;
  }

  /**
   * Checks if a token is valid without consuming it.
   * Returns [teamId, true] if valid and not expired.
   * Returns ['', false] if the token is unknown or expired.
   */
  validate(token: string): [string, boolean] {
    const entry = this.tokens.get(token);
    if (entry === undefined) {
      return ['', false];
    }

    if (this.isExpired(entry.createdAt)) {
      this.tokens.delete(token);
      return ['', false];
    }

    return [entry.teamId, true];
  }

  /**
   * Validates and removes a token in one operation.
   * Returns [teamId, true] if valid; deletes the token.
   * Returns ['', false] if the token is unknown or expired.
   */
  consume(token: string): [string, boolean] {
    const entry = this.tokens.get(token);
    if (entry === undefined) {
      return ['', false];
    }

    if (this.isExpired(entry.createdAt)) {
      this.tokens.delete(token);
      return ['', false];
    }

    this.tokens.delete(token);
    return [entry.teamId, true];
  }

  /**
   * Atomically validates and consumes a token.
   * Equivalent to validate() followed by consume() but in a single operation.
   * Returns [teamId, true] if valid; deletes the token.
   * Returns ['', false] if the token is unknown or expired.
   */
  validateAndConsume(token: string): [string, boolean] {
    return this.consume(token);
  }

  /**
   * Returns the number of unexpired, unused tokens.
   * Useful for testing and monitoring.
   */
  pendingCount(): number {
    let count = 0;
    for (const entry of this.tokens.values()) {
      if (!this.isExpired(entry.createdAt)) {
        count++;
      }
    }
    return count;
  }

  /**
   * Stops the background cleanup interval.
   * Call this when the TokenManager is no longer needed.
   */
  close(): void {
    clearInterval(this.cleanupInterval);
  }

  /**
   * Inserts an already-expired token into the manager.
   * FOR TESTING ONLY — do not call from production code.
   */
  injectExpiredToken(token: string, teamId: string): void {
    const expiredAt = new Date(Date.now() - 2 * TOKEN_TTL_MS);
    this.tokens.set(token, { teamId, createdAt: expiredAt });
  }

  // ---------------------------------------------------------------------------
  // Private helpers
  // ---------------------------------------------------------------------------

  /** Returns true if the token created at `createdAt` has exceeded TOKEN_TTL_MS. */
  private isExpired(createdAt: Date): boolean {
    return Date.now() - createdAt.getTime() > TOKEN_TTL_MS;
  }

  /** Removes all tokens older than TOKEN_TTL_MS from the map. */
  private cleanupExpiredTokens(): void {
    for (const [token, entry] of this.tokens.entries()) {
      if (this.isExpired(entry.createdAt)) {
        this.tokens.delete(token);
      }
    }
  }
}
